gobblin-core/src/main/java/gobblin/instrumented/converter/InstrumentedConverterDecorator.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) 2014 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.instrumented.converter;
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.converter.Converter;
+import gobblin.converter.DataConversionException;
+import gobblin.converter.SchemaConversionException;
+import gobblin.metrics.MetricContext;
+
+
+/**
+ * Decorator that automatically instruments {@link gobblin.converter.Converter}.
+ * Handles already instrumented {@link gobblin.instrumented.converter.InstrumentedConverter} appropriately to
+ * avoid double metric reporting.
+ */
+public class InstrumentedConverterDecorator<SI, SO, DI, DO> extends InstrumentedConverterBase<SI, SO, DI, DO> {
+  private Converter<SI, SO, DI, DO> embeddedConverter;
+  private boolean isEmbeddedInstrumented;
+
+  public InstrumentedConverterDecorator(Converter<SI, SO, DI, DO> converter) {
+    this.embeddedConverter = converter;
+    this.isEmbeddedInstrumented = InstrumentedConverterBase.class.isInstance(converter);
+  }
+
+  @Override
+  public Converter<SI, SO, DI, DO> init(WorkUnitState workUnit) {
+    this.embeddedConverter = this.embeddedConverter.init(workUnit);
+    return super.init(workUnit);
+  }
+
+  @Override
+  public MetricContext getMetricContext() {
+    return this.isEmbeddedInstrumented ?
+        ((InstrumentedConverterBase)embeddedConverter).getMetricContext() :
+        super.getMetricContext();
+  }
+
+  @Override
+  public Iterable<DO> convertRecord(SO outputSchema, DI inputRecord, WorkUnitState workUnit)
```

Comment:
Seems you don't need to override this method since `convertRecordImpl` is also doing `this.embeddedConverter.convertRecord(outputSchema, inputRecord, workUnit)`.